docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,544 @@
+    KubeadmConfigSpec cabpkv1.KubeadmConfigSpec `json:"kubeadmConfigSpec"`
+}
+
+// ControlPlaneStatus defines the observed state of ControlPlane
+type ControlPlaneStatus struct {
+    // Selector is the label selector in string format to avoid introspection
+    // by clients, and is used to provide the CRD-based integration for the
+    // scale subresource and additional integrations for things like kubectl
+    // describe.. The string will be in the same format as the query-param syntax.
+    // More info about label selectors: http://kubernetes.io/docs/user-guide/labels#label-selectors
+    // +optional
+    Selector string `json:"selector,omitempty"`
+
+    // Total number of non-terminated machines targeted by this control plane
+    // (their labels match the selector).
+    // +optional
+    Replicas int32 `json:"replicas,omitempty"`
+
+    // Total number of non-terminated machines targeted by this control plane
+    // that have the desired template spec.
+    // +optional
+    UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
+
+    // Total number of fully running machines, including the number of active
+    // etcd cluster members when running as stacked
+    // +optional
+    ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+
+    // Total number of available machines targeted by this control plane.
+    // This is the same as ReadyReplicas, but exposed to be similar to the
+    // status exposed by MachineDeployments and Deployments
+    // +optional
+    AvailableReplicas int32 `json:"readyReplicas,omitempty"`
+
+    // Total number of unavailable machines targeted by this control plane.
+    // This is the total number of machines that are still required for
+    // the deployment to have 100% available capacity. They may either
+    // be machines that are running but not yet ready or machines
+    // that still have not been created.
+    // +optional
+    UnavailableReplicas int32 `json:"unavailableReplicas,omitempty"`
+
+    // Initialized denotes whether or not the control plane has the
+    // uploaded kubeadm-config configmap.
+    Initialized bool `json:"initialized,omitempty”`
+
+    // Ready denotes that the ControlPlane API Server is ready to
+    // receive requests.
+    Ready bool `json:"ready,omitempty"`
+
+    // ErrorReason indicates that there is a problem reconciling the
+    // state, and will be set to a token value suitable for
+    // programmatic interpretation.
+    // +optional
+    ErrorReason ControlPlaneStatusError `json:"errorReason,omitempty"`
+
+    // ErrorMessage indicates that there is a problem reconciling the
+    // state, and will be set to a descriptive error message.
+    // +optional
+    ErrorMessage *string `json:"errorMessage,omitempty"`
+
+    // Phase represents the current phase of control plane actuation.
+    // E.g. Pending, Initialized, Ready, ScalingUp, ScalingDown, Failed etc.
+    // +optional
+    Phase string `json:"phase,omitempty"`
```

Comment:
The word _phase_ sounds wonky here. Maybe _state_? It works better with well-defined states as well:

```golang
// ControlPlaneState represents the current phase of control plane actuation.
type ControlPlaneState string

const (
	ControlPlanePending     ControlPlaneState = "pending"
	ControlPlaneInitialized ControlPlaneState = "initialized"
	ControlPlaneReady       ControlPlaneState = "ready"
	ControlPlaneScalingUp   ControlPlaneState = "scaling-up"
	ControlPlaneScalingDown ControlPlaneState = "scaling-down"
	ControlPlaneFailed      ControlPlaneState = "failed"
)
```

However, the Kubernetes API conventions indicates [_phase_ is being deprecated in favor of conditions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) anyway:

> Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. The pattern of using phase is deprecated. Newer API types should use conditions instead. 

How would we rewrite the above as conditions?

Reply:
> The word phase sounds wonky here. Maybe state?

We've discussed this during the v1alpha2 planning, `phase` is the preferred word to describe the state the object is in,  I'd prefer if we keep things consistent between our types.

 > How would we rewrite the above as conditions?

This is a larger conversation on how we define and describe a state machine. @timothysc is thinking about this problem already, I'd suggest to move this topic to an issue though.

Reply:
This is something that we discussed earlier on the Google doc and decided to stick with Phase to align with the other Cluster API components for now, and we should attempt to address this holistically in the future. /cc @timothysc

Reply:
Isn't it exactly the time to adopt recommendations rather than defer, defer, defer? Just because it's hard doesn't mean it's wrong. From what I read _conditions_ are the recommended approach, and it seems like we should figure out how to use them since this is net-new work.

Reply:
We also have this note that says Conditions themselves are being deprecated https://github.com/detiber/cluster-api/blob/master/docs/proposals/machine-api-proposal.md#conditions

Reply:
Believe this is the history of the statements: https://github.com/kubernetes/kubernetes/issues/51594#issuecomment-334646068

It's a case not to base things not on Deployments in the core API, but on the newer StatefulSets and DaemonSets instead.

Reply:
Considering the contention this topic brings, I'd like to defer this to the issue created to track this discussion more broadly: https://github.com/kubernetes-sigs/cluster-api/issues/1658